community/cypher/cypher/src/test/scala/org/neo4j/cypher/VarLengthAcceptanceTest.scala

```diff
@@ -0,0 +1,384 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher
+
+import org.neo4j.cypher.internal.compiler.v2_2.commands.NoneInCollection
+import org.neo4j.cypher.internal.compiler.v2_2.executionplan.InternalExecutionResult
+import org.neo4j.cypher.internal.compiler.v2_2.planDescription.InternalPlanDescription
+import org.neo4j.cypher.internal.compiler.v2_2.planDescription.InternalPlanDescription.Arguments.LegacyExpression
+import org.neo4j.graphdb.{Direction, Node}
+import org.neo4j.graphdb.Direction._
+import org.scalatest.matchers.{MatchResult, Matcher}
+
+import scala.collection.mutable
+
+class VarLengthAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsTestSupport with NewPlannerTestSupport {
+
+  /*
+  This tree model generator will generate a binary tree, starting with a single root(named "n0").
+  'maxNodeDepth' refers to the node depth, so a value of 4 will generate a root, 2 children, 4 grandchildren
+   and 8 grandchildren (so the depth of the tree is 3)
+  */
+  private def makeTreeModel(maxNodeDepth: Int, directions: Seq[Direction] = Seq()) = {
+
+    val nodes = mutable.Map[String, Node]()
+    Range(0, maxNodeDepth).foreach { depth =>
+      val width = Range(0, depth).fold(1) { (acc, n) =>
+        acc * 2
+      }
+      Range(0, width).foreach { index =>
+        val inum = "0" * width + index.toBinaryString
+        val name = "n" + inum.substring(inum.length - (depth + 1), inum.length)
+        val parentName = name.substring(0, name.length - 1)
+        nodes(name) = createNode(Map("id" -> name))
+        if (nodes.isDefinedAt(parentName)) {
+          val dir = if (directions.length >= depth) directions(depth - 1) else OUTGOING
+          dir match {
+            case OUTGOING =>
+              relate(nodes(parentName), nodes(name), "LIKES")
+            case INCOMING =>
+              relate(nodes(name), nodes(parentName), "LIKES")
+            case _ =>
+              throw new UnsupportedOperationException("Only accept INCOMING and OUTGOING")
+          }
+        }
+      }
+    }
+    nodes.toMap
+  }
+
+  //The root will be generation 1
+  private def generation(gen: Int)(implicit nodes: Map[String, Node]): Set[Node] =
+    nodes.filter { elem =>
+      elem._1.length == gen + 1
+    }.values.toSet match {
+      case s: Set[Node] if s.nonEmpty => s
+      case _ => throw new UnsupportedOperationException("Unexpectedly empty test set")
+    }
```

Comment:
Perhaps it would be a good idea to move all this utility to the bottom of the class, as I've gotten the idea that we tend to do that for other test classes.